scripts/configure.utils.py

```diff
@@ -0,0 +1,157 @@
+            return
+        self.args[arg] = json.dumps(value if type != '[' else value.split() if list else [value])
+
+    def gn_arg(self, name, prefixes=(), type=None):
+        """Finds the GN argument corresponding to a parameter name"""
+        arg = name.translate(str.maketrans('-', '_'))
+        candidates = [p + arg for p in (('',) + prefixes) if (p + arg) in self.gn_args]
+        preferred = [c for c in candidates if self.gn_args[c] == type] if type else []
+        if not (match := next(iter(preferred + candidates), None)):
+            info("Warning: Project has no build arg '%s'" % arg)
+        return match
+
+    def process_triplet_parameter(self, name, value):
+        if value is None:
+            fail("Project option --%s requires an argument" % name)
+        triplet = value.split('-')
+        if len(triplet) not in (2, 3, 4):
+            fail("Project option --%s argument must be a cpu-vendor-os[-abi] triplet" % name)
+        prefix = 'host_' if name == 'build' else 'target_'  # "host" has different meanings in GNU and GN!
+        self.add_arg(prefix + 'cpu', triplet[0])
+        self.add_arg(prefix + 'os', triplet[1 if len(triplet) == 2 else 2])
+
+    def process_enable_parameter(self, name, value):
+        if not (arg := self.gn_arg(name[len('enable-'):], self.BOOL_ARG_PREFIXES, 'b')):
+            return
+        if self.gn_args[arg] != 'b':
+            fail("Project build arg '%s' is not a boolean" % arg)
+        if value != 'no' and value is not None:
+            fail("Invalid argument for --%s, must be absent or 'no'" % name)
+        self.add_arg(arg, value is None)
+
+    def process_generic_parameter(self, name, value):
+        if not (arg := self.gn_arg(name, self.GENERIC_ARG_PREFIXES)):
+            return
+        if self.gn_args[arg] == 'b':
+            fail("Project build arg '%s' is a boolean, use --enable-..." % arg)
+        if value is None:
+            fail("Project option --%s requires an argument" % name)
+        self.add_arg(arg, value)
+
+    def process_parameter(self, name, value):
+        if name in ('build', 'host', 'target'):
+            self.process_triplet_parameter(name, value)
+        elif name.startswith('enable-'):
+            self.process_enable_parameter(name, value)
+        else:
+            self.process_generic_parameter(name, value)
+
+    def process_parameters(self, argv):
+        """Process GNU-style configure command line parameters"""
+        for arg in argv:
+            if not (m := re.fullmatch(r'--([a-z][a-z0-9-]*)(?:=(.*))?', arg)):
+                fail("Invalid argument: '%s'" % arg)
+            self.process_parameter(m.group(1), m.group(2))
+
+
+def info(message):
+    print(message, file=sys.stderr)
+
+
+def fail(message):
+    info("Error: " + message)
+    sys.exit(1)
+
+
+if __name__ == '__main__' and len(sys.argv) >= 2:
+    [_, func, *args] = sys.argv
+    globals()[func](*args)
```

Comment:
this is tricky code. Avoid tricky code.

Reply:
Added a comment and made the code unconditional since it's only called from `configure`.